src/core/drive/preloader.ts

```diff
@@ -0,0 +1,48 @@
+import { expandURL } from "../url"
+import { Navigator } from "./navigator"
+import { PageSnapshot } from "./page_snapshot"
+import { SnapshotCache } from "./snapshot_cache"
+
+export interface PreloaderDelegate {
+  readonly navigator: Navigator
+}
+
+export class Preloader {
+  readonly delegate: PreloaderDelegate
+  readonly element: Element
+
+  constructor(delegate: PreloaderDelegate, element: Element) {
+    this.delegate = delegate
+    this.element = element
+  }
+
+  get snapshotCache(): SnapshotCache {
+    return this.delegate.navigator.view.snapshotCache
+  }
+
+  start() {
+    this.preloadOnLoadLinksForView(this.element)
+  }
+
+  preloadOnLoadLinksForView(element: Element) {
+    const links = element.querySelectorAll('a[data-turbo-preload="load"]')
+
+    for (const link of links) {
+      this.preloadURL(link)
+    }
+  }
+
+  preloadURL(link: Element) {
+    const url = expandURL(link.getAttribute("href") || "")
+
+    if (this.snapshotCache.has(url)) {
+      return
+    }
+
+    fetch(url.toString(), { headers: { 'VND.PREFETCH': 'true', 'Accept': 'text/html' } })
+      .then(res => res.text())
+      .then(responseText => PageSnapshot.fromHTMLString(responseText))
+      .then(snapshot => this.snapshotCache.put(url, snapshot))
+      .catch(() => {})
```

Comment:
This codebase utilizes [`async`/`await` syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/async_function), which enables this to be implemented without chaining:

```js
const response = await fetch(url.toString(), { headers: { 'VND.PREFETCH': 'true', 'Accept': 'text/html' } })
const responseText = await response.text()
const snapshot = PageSnapshot.fromHTMLString(responseText)

this.snapshotCache.put(url, snapshot)
```

Reply:
Yes that does seem to be the convention. I will tidy this up 👍. Thank you.

Reply:
cfd830d0df4c46a4383e620117005e113dd77472